### DISCUSSION TITLE
Welcome to HammerDB Discussions!

### DISCUSSION THREAD

**Discussion by sm-shaw** [TPC-Council/HammerDB#181] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by nhariba2** [comment]
Am trying to build the schema in hammerdb but its taking time to creating rows on tables its keeps on generating nearly 1lakhs records. May i know how to reduce number of records creation, due to this issue am unable to proceed on hammerdb. Its taking more than 5hrs also still records keep on generating. May i know the values of virtual users and warehouse, etc to be entered? Here am help less. I have given varies values virtual user and warehouse still same issues taking more time to complete schema build. I have entered 1wareshouse and one virtual user for 8gb ram and 2core sql machine keeps on generating the rows in tables... Also let me know how many records will be generate on default while build schema on each tables?

**Comment by jwbarker736** [comment]
All,
I am working on a project that uses HammerDB to simulate workloads on databases and tracks the communication between the HammerDB server and the database.  Currently, I am attempting to simulate a 3-tier (web, application and database) application in which the HammmerDB server simulates the application tier of the 3-tier application.

**Comment by jwbarker736** [comment]
All,

I apologize if I missed this some where in the discussions, but I combed the archive and could not find what I’m looking for.

In the 08Jul19 HammerDB Blog, Using HammerDB as a Web Service the author goes into great detail regarding the syntax in the restchk.tcl script to test the Web Service interface to the HammerDB server.  And I’ve got the script functioning correctly on my platform.  However, following the discussion regarding the restchk.tcl script, the author is very brief regarding the development of scripts to build schemas and run workloads (i.e., When the environment is configured you can build schemas and run workloads with the same commands used for the CLI, for example:).  The examples are as follows:

set res [rest::post http://localhost:8080/buildschema "" ]
set res [rest::post http://localhost:8080/vurun "" ]

The author concludes with “The key difference is that output is now stored in a job format that can be retrieved at a later point. An example is shown where a build generates a jobid.” And a third example:

{"success": {"message": "Building 5 Warehouses with 6 Virtual Users, 5 active + 1 Monitor VU(dict value num_vu is set to 5): JOBID=5D23464E58D203E273738333"}}

That is then used to query the status of the build.  The output is stored in a SQLite database that by default runs in in-memory mode that is not persistent over restarts, however it can be configured to use a permanent database.
It is not intended for the web service to replace the GUI or CLI environments that still prove the most popular way to run HammerDB, however the aim is to provide an additional way that HammerDB can be integrated into tests, in cloud environments particular.

Here is my ask, as I said above, I have the restchk.tcl script functioning on my platform.  I understand how to do what I am attempting to accomplish via the GUI and CLI, but I am struggling with the syntax for buildschema and vurun scripts.  
1.	Does anyone have an example buildschema and/or vurun script, ideally written in tcl, they would be willing to share?

I am unclear on how you obtain the JOBID in order to query the status of a build.
2.	Can anyone explain how to obtain the JOBID in order to query the status of a build?

Any example scripts or other insights would be greatly appreciated.

**Comment by sm-shaw** [comment]
Yes the information you need is in the documentation:

Running Jobs: https://www.hammerdb.com/docs/ch10s04.html
Querying Jobs: https://www.hammerdb.com/docs/ch10s05.html

So the JOBID is printed as the result of a successful build or run (ie it successfully started) 

BUILD
`{"success": {"message": "Building 10 Warehouses with 6 Virtual Users, 5 active + 1 Monitor VU(dict value num_vu is set to 5): JOBID=5D1F4CA858CE03E213431323"}}`

RUN
`{"success": {"message": "Running Virtual Users: JOBID=5D1F4FF558CE03E223730313"}}`

You then query the JOBID to get the output that in the CLI is printed to stdout.

QUERY
`set res [rest::get http://localhost:8080/jobs?jobid=5D1F4FF558CE03E223730313 "" ]`

When you start the webservice the home page lists the available commands. It is very similar to the CLI except for the interface.

**Comment by xpfb** [comment]
Hello  - wanting to use "HammerDB" for some internal "cloud" testing. Have had to build hammerdb RHEL 7.x  from scratch - corporate has everything blocked and firewalled to death. Just e-mailed everything - vnc and hammerdb4.1, then have compiled and link-loaded 8.6.x tcl/tk/tkblt/tksvg(.7) - am close but am throwing this error at "./hammerdb" invocation - splash screen shows, then...
_While loading component file "gentheme.tcl"...
can't find package tksvg while executing
"package require tksvg" (procedure "initscaletheme" line 7) invoked from within "initscaletheme $theme $pixelsperpoint"
(file ".src/generic/gettheme.tcl" line 479) invoked from within "source [ file join $UserDefaultDir src generic $f ]"_

Guidance much appreciated - thank you.
**(BTW- have used same for 30 yrs - there's nothing like it for pushing/stressing your (ORA) DB environments.)**

**Reply by sm-shaw** [reply]
Hi,
Yes since v4.0 there is a new Graphical interface that scales for UHD displays - this uses TKSVG that is in the lib directory so it looks like you are missing this directory. Alternatively if you don't need scalable graphics then try setting the graphics to fixed as described here https://www.hammerdb.com/docs/ch01s10.html this will then use the previous fixed display it will not scale but will be faster over VNC.

**Comment by xpfb** [comment]
Am now running into this "theme" issue at 4-1,"hammerDB" invocation
While loading component file "gentheme.tcl"..._
_can't find package ttk::theme::clearlooks while executing
"package require ttk:: theme::$theme"
 (procedure "ttk::setTheme" line 4) invoked from within
"ttk::setTheme $theme"
 (procedure "initfixedthem" line 17) invoked from within
"initfixedtheme $theme"
(file "./scr/generic/gentheme.tcl" line 502) invoked from within
"source [ file join $UserDefaultDir src generic $f ]"

Still haven't got "tksvg" sorted - did try the generic.xml <theme> settings though and now this...
Cheers

**Comment by sm-shaw** [comment]
clearlooks is a non-compiled theme package in the lib directory.  This is the theme used on Linux when not using SVG.
You can just copy this over from the lib directory of any installation.
tksvg is here https://github.com/oehhar/tksvg - note hammerdb uses version 0.5. There were issues with 0.6 so it is important to get the right one. 
This is why all of the packages are pre-compiled so it 'just works' I've done all this work before for both Linux and Windows.

**Comment by tieonster** [comment]
Hi all,

I've been running a small test using HammerDB 4.7 with Oracle 19, and have been obtaining 0 TPM. NOPM seems to be displaying a value.  I understand that this issue has been discussed here and has been fixed quite some time ago (#[189](https://github.com/TPC-Council/HammerDB/issues/189)) 

However, I am still facing the same problem and getting 0 TPM. Can I check how I can go about resolving this issue on my end? Any help would be greatly appreciated!

**Comment by laharinakka** [comment]
Hi I am trying to generate data in Hammer DB for Oracle and receiving the below error could you please help me 
Error in Virtual User 1: Error: wrong # args: should be "do_tpcc system_user system_password instance count_ware tpcc_user tpcc_pass tpcc_def_tab tpcc_ol_tab tpcc_def_temp partition timesten hash_clusters num_vu"

**Reply by sm-shaw** [reply]
Most likely special characters in a field such as a password. In this case wrap the entry with curly brackets {} to prevent interpretation as special characters

**Comment by ananthu-dev** [comment]
Hello Steve, 
I'm trying to understand how to send http requests to hammerdb web service. I was looking at the documentation here: https://www.hammerdb.com/blog/uncategorized/using-hammerdb-as-a-web-service/. When I navigate to port 8080 via Chrome, I am greeted with the jobs page instead of api documentation page. 
Also when I send api requests, for example, http://localhost:8080/bm as a test I get the jobs html page back. I am using postman to send get requests. 
Am I missing any headers? I thought I would get a json with the set benchmark as reply.

**Reply by sm-shaw** [reply]
Hi Yes, this functionality has been deprecated as described here https://github.com/TPC-Council/HammerDB/issues/420 - the article you reference is an older blog post - we started to add REST based functionality, however after discussion in the TPC-OSS subcommittee it was simply not being used (see the bug referenced in the issue above) and development time was better spent elsewhere. 

Instead, the web service changed into a read-only results repository for workloads run by the GUI & CLI (you can also run the GUI as a web application in a browser https://www.hammerdb.com/blog/uncategorized/deploy-hammerdb-as-a-web-application-with-cloudtk/) 

Building an interactive web service with the browser interface as we started would be straightforward, starting with the functionality that was deprecated, for now there are no plans to do so from the core team, however this could be contributed if there was a desire to do this.
